File: examples/canonical-analytics/python/src/analytics_workbench/analytics/distributions.py
from dataclasses import dataclass
from ..data.models import LatencySample, Service
from ..filters import FilterState
import numpy as np

@dataclass
class LatencyHistogram:
    values: list[float]
    bin_size: float
    p50: float
    p95: float
    p99: float

@dataclass
class LatencyByService:
    services: list[str]
    p50s: list[float]
    p95s: list[float]
    p99s: list[float]

def _filter_samples(samples: list[LatencySample], valid_services: set, filters: FilterState):
    filtered = []
    for s in samples:
        d = s.timestamp[:10]
        if filters.date_from and d < filters.date_from.isoformat(): continue
        if filters.date_to and d > filters.date_to.isoformat(): continue
        
        if s.service not in valid_services: continue
        if filters.services and s.service not in filters.services: continue
        
        filtered.append({"service": s.service, "latency_ms": s.latency_ms})
    return filtered

def aggregate_latency_histogram(samples: list[LatencySample], filters: FilterState, services: list[Service] = None) -> LatencyHistogram:
    if services is not None:
        valid_services = {s.name for s in services}
        filtered = _filter_samples(samples, valid_services, filters)
    else:
        filtered = []
        for s in samples:
            d = s.timestamp[:10]
            if filters.date_from and d < filters.date_from.isoformat(): continue
            if filters.date_to and d > filters.date_to.isoformat(): continue
            if filters.services and s.service not in filters.services: continue
            filtered.append({"service": s.service, "latency_ms": s.latency_ms})
    
    if not filtered:
        return LatencyHistogram(values=[], bin_size=10.0, p50=0.0, p95=0.0, p99=0.0)
        
    values = [f["latency_ms"] for f in filtered]
    
    return LatencyHistogram(
        values=values,
        bin_size=10.0,
        p50=float(np.percentile(values, 50)),
        p95=float(np.percentile(values, 95)),
        p99=float(np.percentile(values, 99))
    )

def aggregate_latency_by_service(samples: list[LatencySample], services: list[Service], filters: FilterState) -> LatencyByService:
    import pandas as pd
    valid_services = {s.name for s in services}
    filtered = _filter_samples(samples, valid_services, filters)
    
    if not filtered:
        return LatencyByService(services=[], p50s=[], p95s=[], p99s=[])
        
    df = pd.DataFrame(filtered)
    agg = df.groupby("service").agg(
        p50=("latency_ms", lambda x: np.percentile(x, 50)),
        p95=("latency_ms", lambda x: np.percentile(x, 95)),
        p99=("latency_ms", lambda x: np.percentile(x, 99))
    )
    agg = agg.sort_values("p50", ascending=False)
    
    return LatencyByService(
        services=agg.index.tolist(),
        p50s=agg["p50"].tolist(),
        p95s=agg["p95"].tolist(),
        p99s=agg["p99"].tolist()
    )

File: examples/canonical-analytics/go/internal/analytics/distributions.go
package analytics

import (
	"sort"

	"analytics-workbench-go/internal/data"
	"analytics-workbench-go/internal/filters"
)

type LatencyHistogram struct {
	Values []float64
	P50    float64
	P95    float64
	P99    float64
}

type LatencyByService struct {
	Services []string
	P50s     []float64
	P95s     []float64
	P99s     []float64
}

func filterSamples(samples []data.LatencySample, services []data.Service, f filters.FilterState) []data.LatencySample {
	validServices := make(map[string]bool)
	if len(services) > 0 {
		for _, s := range services {
			validServices[s.Name] = true
		}
	} else {
		for _, s := range samples {
			validServices[s.Service] = true
		}
	}

	var filtered []data.LatencySample
	for _, s := range samples {
		if !validServices[s.Service] {
			continue
		}
		if !f.ContainsService(s.Service) {
			continue
		}
		t := parseDate(s.Timestamp)
		if !f.InDateRange(t) {
			continue
		}
		// Notice: Python doesn't filter environment in latency? Python code does not apply env filter in _filter_samples.
		filtered = append(filtered, s)
	}
	return filtered
}

func AggregateLatencyHistogram(samples []data.LatencySample, services []data.Service, f filters.FilterState) LatencyHistogram {
	filtered := filterSamples(samples, services, f)
	if len(filtered) == 0 {
		return LatencyHistogram{}
	}

	values := make([]float64, len(filtered))
	for i, s := range filtered {
		values[i] = s.LatencyMs
	}

	// copy values to compute percentiles since percentile sorts in place
	vCopy := append([]float64(nil), values...)
	
	return LatencyHistogram{
		Values: values,
		P50:    percentile(vCopy, 50),
		P95:    percentile(vCopy, 95),
		P99:    percentile(vCopy, 99),
	}
}

func AggregateLatencyByService(samples []data.LatencySample, services []data.Service, f filters.FilterState) LatencyByService {
	filtered := filterSamples(samples, services, f)
	if len(filtered) == 0 {
		return LatencyByService{}
	}

	bySrv := make(map[string][]float64)
	for _, s := range filtered {
		bySrv[s.Service] = append(bySrv[s.Service], s.LatencyMs)
	}

	type srvP struct {
		Name string
		P50  float64
		P95  float64
		P99  float64
	}

	var srvs []srvP
	for name, vals := range bySrv {
		srvs = append(srvs, srvP{
			Name: name,
			P50:  percentile(append([]float64(nil), vals...), 50),
			P95:  percentile(append([]float64(nil), vals...), 95),
			P99:  percentile(append([]float64(nil), vals...), 99),
		})
	}

	sort.Slice(srvs, func(i, j int) bool {
		return srvs[i].P50 > srvs[j].P50
	})

	res := LatencyByService{
		Services: make([]string, len(srvs)),
		P50s:     make([]float64, len(srvs)),
		P95s:     make([]float64, len(srvs)),
		P99s:     make([]float64, len(srvs)),
	}

	for i, s := range srvs {
		res.Services[i] = s.Name
		res.P50s[i] = s.P50
		res.P95s[i] = s.P95
		res.P99s[i] = s.P99
	}

	return res
}
